### PR TITLE
Support DiffusionGemma (google/diffusiongemma-26B-A4B-it)

### PR DESCRIPTION
**Pull request by rwang5203**
## Motivation

Day-0 support for DiffusionGemma (google/diffusiongemma-26B-A4B-it), an encoder-decoder, uniform-state (renoising) block-diffusion model (26B-A4B MoE) that renoises the canvas from random tokens rather than an absorbing mask state, with a Gemma4 vision tower. Internal reference PR.

## Modifications

The EntropyBound sampler denoises a fixed-length canvas over max_denoising_steps reverse steps with self-conditioning: it accepts the lowest-entropy canvas positions within entropy_bound, re-noises the complement, applies a linear temperature schedule and StableAndConfident stopping, and emits the greedy argmax of the processed logits. The Gemma4 vision tower injects image soft tokens into the encoder input embeddings only, so the decoder canvas never sees image tokens.

- `models/gemma4_diffusion.py`: the encoder (causal context) / decoder (bidirectional canvas) model, with self-conditioning, the MoE gate routed on the raw residual, and a first-rank vision tower plus multimodal embedder whose image features feed the encoder pass through `general_mm_embed_routine`.
- `dllm/algorithm/gemma4_renoise.py`: the EntropyBound accept/renoise sampler. Per-request stop and rank-shared seeded RNG keep batched and TP runs correct.
- `multimodal/processors/diffusion_gemma.py`: an image-only processor reusing the checkpoint's Gemma4 image processor.
- A uniform-state path through the dLLM scheduler (`is_uniform`). Each committed canvas is re-encoded causally before the next round, since the KV the decoder leaves behind is its transient denoising KV and reusing it as context degrades long generations. The transient canvas KV is also kept out of the radix tree on finish.
- dLLM pool robustness: admissions are capped by the actual allocator headroom (full and SWA pools), a saturated pool aborts the least-progressed request with a retryable 503 instead of crashing the scheduler (victim order mirrors `retract_decode`, with new-request back-pressure while the pool is tight), and a request whose next canvas would overrun its `req_to_token` row finishes with `length`.
- Auto-config: `--dllm-algorithm Gemma4Renoise` forces triton, eager mode, and unchunked prefill (head_dim 512, bidirectional canvas), so a default launch works with no extra flags.
- `logprobs`, penalties, `logit_bias`, and grammar / structured output are rejected with a 400 (the renoise schedule cannot honor them).

## Accuracy Tests

google/diffusiongemma-26B-A4B-it, full test splits, no sampling. MCQ via greedy generate-and-parse, math via numeric last-number matching with boxed-answer extraction plus sympy, HumanEval via fenced-code extraction. Every item is scored (no failed-request exclusions). MMLU, ARC-Challenge, and MATH-500 are the mean of two independent server launches (the renoise RNG reseeds per launch, observed run-to-run spreads 0.3, 1.8, and 3.0 points respectively). Non-thinking tables were measured on an earlier build of this branch. Later changes were verified benchmark-neutral with paired tests and launch means, and non-thinking generations rarely exceed the 1024-token window where the only output-affecting change applies, so treat these as a serving-correctness baseline that may slightly understate the final branch.

| Text benchmark | Score |
|---|---|
| GSM8K | 95.4% |
| ARC-Challenge | 91.6% |
| HumanEval | 92.7% pass@1 |
| MMLU | 76.2% |
| GSM-Symbolic | 92.2% |
| MATH-500 | 72.1% |
| GPQA-main | 59.2% |

Multimodal, full standard split per task. MMMU and MMStar as multiple-choice, MathVista testmini, DocVQA by ANLS, ChartQA by relaxed accuracy, AI2D as multiple-choice.

| Multimodal benchmark | Score |
|---|---|
| MMMU (val, MC) | 64.9% |
| MathVista (testmini) | 68.4% |
| DocVQA (val) | 85.9% |
| ChartQA (test) | 61.7% |
| AI2D (test) | 78.7% |
| MMStar (val) | 65.9% |

Thinking mode, enabled via the `<|think|>` chat-template trigger, with answers parsed after the thinking channel closes. Per dataset:

- AIME-2026: budget 32768, mean of 7 independent server launches (the renoise RNG reseeds per launch, observed spread of plus or minus 9 points on the 30-item set).
- GPQA-Diamond: budget 16384, mean of 10 independent launches.
- MMMLU: budget 4096, macro-average over the 14 per-language test sets on a balanced 200-per-language sample.
- MMMU-Pro: budget 8192, vision split (question embedded in a screenshot), matching the setting the model card reports. The easier standard 10-option split scores 65.0%.
- MMLU-Pro: budget 8192, full test split.
- LiveCodeBench v6: budget 24576, full release_v6 set (1055 problems), scored with the official lcb_runner pass@1 evaluator.

| Thinking benchmark | This PR | Model card |
|---|---|---|
| AIME-2026 (no tools) | 72.4% (7 launches) | 69.1% |
| GPQA-Diamond | 72.1% (10 launches) | 73.2% |
| MMMLU (14-language macro) | 83.1% | 81.5% |
| MMMU-Pro (vision) | 55.2% | 54.3% |
| MMLU-Pro | 79.5% | 77.6% |
| LiveCodeBench v6 | 76.9% | 69.1% |

## Speed Tests and Profiling

Not benchmarked for speed. This adds a new model and the Gemma4Renoise sampler.

## Checklist

- [ ] Format the code with pre-commit (run before merging).
- [ ] Unit tests: none added (internal reference PR).
- [x] Documentation updated (`docs/supported_models/text_generation/diffusion_language_models.md`).
- [x] Accuracy results provided (above). Speed benchmark not applicable.
- [ ] Code style guidance (covered by pre-commit).

---
### CI States

Latest PR Test (Base): :x: [Run #27502196936](https://github.com/sgl-project/sglang/actions/runs/27502196936)
Latest PR Test (Extra): :x: [Run #27502196825](https://github.com/sgl-project/sglang/actions/runs/27502196825)

### CODE REVIEW

File: python/sglang/srt/dllm/algorithm/gemma4_renoise.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            else:
+                self_cond = torch.where(
+                    done.view(bs, 1, 1), self_cond.view(bs, L, -1), logits
+                ).reshape(bs * L, -1)
+
+            done |= self._stop(history, argmax, token_entropy) & (~done)
+            if bool(done.all()):
+                break
+
+        next_token_ids = [argmax[i] for i in range(bs)]
```
[the remark this answers is not in the document]
Comment (mickqian) [reply python/sglang/srt/dllm/algorithm/gemma4_renoise.py:164]: this is worth noticing, and make sure do `argmax.cpu()` to avoid multiple D2H

[the remark this answers is not in the document]
Comment (rwang5203) [reply python/sglang/srt/dllm/algorithm/gemma4_renoise.py:164]: Done in b70982b, single argmax.cpu() per round before the split, so the per-request tolist() reads host memory.

File: python/sglang/srt/server_args.py [hunk 2]
```diff
@@ -4528,6 +4528,24 @@ def _handle_deterministic_inference(self):
...
+                logger.warning(
+                    "Attention backend forced to triton for DiffusionGemma "
+                    "(head_dim 512 exceeds the flashinfer/fa3 cap)."
+                )
+                self.attention_backend = "triton"
+                self.prefill_attention_backend = None
+                self.decode_attention_backend = None
+            self.cuda_graph_config.decode.backend = Backend.DISABLED
+            self.cuda_graph_config.prefill.backend = Backend.DISABLED
+            self.chunked_prefill_size = -1
```
Comment (kpham-sgl) [inline python/sglang/srt/server_args.py:4549]: `trtllm_mha` recently supports headdim 512, Gemma 4 has been using it since. Can you give it a try for DiffusionGemma?

**Review by kpham-sgl — COMMENTED** [review]
Can you resolve conflicts again 😅

File: python/sglang/srt/server_args.py (hunk 2 above)
Comment (rwang5203) [reply python/sglang/srt/server_args.py:4549]: `head_dim` 512 works, but `trtllm_mha` crashes on DiffusionGemma's bidirectional denoise because flashinfer rejects causal=False with window_left>=0 on the 25 sliding-window layers. It can't replace triton here, so the options are a hybrid (trtllm_mha for the 5 full layers, triton for the rest) or waiting for flashinfer to support non-causal sliding-window.

Comment (rwang5203) [reply python/sglang/srt/server_args.py:4549]: Conflicts resolved.
